File: python/src/siphon/orchestrator.py
"""Core decision logic: cache check, size threshold, native upload, inline fallback."""

from __future__ import annotations

import logging
from typing import Iterator

from siphon.base64_stream import encode_chunks_to_base64
from siphon.cache import UploadCache
from siphon.hashing import TeeHasher, hash_chunks
from siphon.providers import Provider
from siphon.sources import MediaSource

logger = logging.getLogger("siphon")

DEFAULT_SIZE_THRESHOLD = 262144


def _inline_block(provider: Provider, chunks: Iterator[bytes], mime_type: str) -> dict:
    b64 = encode_chunks_to_base64(chunks)
    return provider.build_inline_block(b64, mime_type)


def encode_media_sync(
    provider: Provider,
    provider_name: str,
    source: MediaSource,
    cache: UploadCache,
    size_threshold: int = DEFAULT_SIZE_THRESHOLD,
    allow_inline_fallback: bool = True,
) -> dict:
    supports_type = provider.supports_media_type(source.mime_type)

    if source.seekable:
        content_hash = hash_chunks(source.chunks())
        cached = cache.get(provider_name, content_hash)
        if cached is not None:
            return provider.build_reference_block(cached, source.mime_type)

        should_go_inline = not supports_type or (
            source.size is not None and source.size < size_threshold
        )
        if should_go_inline:
            return _inline_block(provider, source.chunks(), source.mime_type)

        try:
            ref = provider.upload(source.chunks(), source.mime_type)
        except Exception:
            if not allow_inline_fallback:
                raise
            logger.warning(
                "siphon: native upload to %r failed; falling back to inline base64 "
                "(this reintroduces the size/memory overhead Siphon avoids).",
                provider_name,
                exc_info=True,
            )
            return _inline_block(provider, source.chunks(), source.mime_type)

        cache.put(provider_name, content_hash, ref)
        return provider.build_reference_block(ref, source.mime_type)

    # Non-seekable: no pre-hash, no proactive size-threshold inline path (size may
    # be unknown). Always attempt native upload if the type is supported; hash is
    # computed as a side effect of the single read pass and cached only on success.
    if not supports_type:
        raise ValueError(
            f"{provider_name!r} does not support media type {source.mime_type!r} "
            "and the source is non-seekable, so no inline fallback is possible "
            "(inline encoding would require re-reading the stream from the start)."
        )

    hasher = TeeHasher()

    def _tee() -> Iterator[bytes]:
        for chunk in source.chunks():
            hasher.update(chunk)
            yield chunk

    ref = provider.upload(_tee(), source.mime_type)
    cache.put(provider_name, hasher.hexdigest(), ref)
    return provider.build_reference_block(ref, source.mime_type)

File: typescript/src/orchestrator.ts
import { encodeChunksToBase64 } from "./base64Stream.js";
import { UploadCache } from "./cache.js";
import { TeeHasher, hashChunks } from "./hashing.js";
import type { Provider } from "./providers/types.js";
import type { MediaSource } from "./sources.js";

export const DEFAULT_SIZE_THRESHOLD = 262144;

async function inlineBlock(
  provider: Provider,
  chunks: AsyncIterable<Buffer>,
  mimeType: string
): Promise<Record<string, unknown>> {
  const base64Data = await encodeChunksToBase64(chunks);
  return provider.buildInlineBlock(base64Data, mimeType);
}

export async function encodeMediaCore(
  provider: Provider,
  providerName: string,
  source: MediaSource,
  cache: UploadCache,
  sizeThreshold: number = DEFAULT_SIZE_THRESHOLD,
  allowInlineFallback = true
): Promise<Record<string, unknown>> {
  const supportsType = provider.supportsMediaType(source.mimeType);

  if (source.seekable) {
    const contentHash = await hashChunks(source.chunks());
    const cached = cache.get(providerName, contentHash);
    if (cached) {
      return provider.buildReferenceBlock(cached, source.mimeType);
    }

    const shouldGoInline =
      !supportsType || (source.size !== null && source.size < sizeThreshold);
    if (shouldGoInline) {
      return inlineBlock(provider, source.chunks(), source.mimeType);
    }

    try {
      const ref = await provider.upload(source.chunks(), source.mimeType);
      cache.put(providerName, contentHash, ref);
      return provider.buildReferenceBlock(ref, source.mimeType);
    } catch (err) {
      console.warn(
        `[siphon] native upload to "${providerName}" failed; falling back to inline base64 ` +
          `(this reintroduces the size/memory overhead Siphon avoids).`,
        err
      );
      if (!allowInlineFallback) throw err;
      return inlineBlock(provider, source.chunks(), source.mimeType);
    }
  }

  if (!supportsType) {
    throw new Error(
      `"${providerName}" does not support media type "${source.mimeType}" and the source is ` +
        `non-seekable, so no inline fallback is possible.`
    );
  }

  const hasher = new TeeHasher();
  async function* tee(): AsyncIterable<Buffer> {
    for await (const chunk of source.chunks()) {
      hasher.update(chunk);
      yield chunk;
    }
  }

  const ref = await provider.upload(tee(), source.mimeType);
  cache.put(providerName, hasher.hexDigest(), ref);
  return provider.buildReferenceBlock(ref, source.mimeType);
}
